### ISSUE TITLE
Consistently Getting TaskCancelledException When ThrowOperationCanceledOnCancellation = true

### ISSUE THREAD

**Issue by thefringeninja** [grpc/grpc-dotnet#731]
We are occasionally having failing tests in CI where we get back `---- Grpc.Core.RpcException : Status(StatusCode=Cancelled, Detail="")` instead of the `GrpcException` we expect. So, I thought I would set `ThrowOperationCanceledOnCancellation = true` on the channel just to see what would happen. Now these tests fail consistently, even locally. I've traced it down to a client interceptor we are using:

```
	public class TypedExceptionInterceptor : Interceptor {
		public override AsyncServerStreamingCall<TResponse> AsyncServerStreamingCall<TRequest, TResponse>(
			TRequest request,
			ClientInterceptorContext<TRequest, TResponse> context,
			AsyncServerStreamingCallContinuation<TRequest, TResponse> continuation) {
			var response = continuation(request, context);

			return new AsyncServerStreamingCall<TResponse>(new AsyncStreamReader<TResponse>(response.ResponseStream),
				response.ResponseHeadersAsync, response.GetStatus, response.GetTrailers, response.Dispose);
		}
		class AsyncStreamReader<TResponse> : IAsyncStreamReader<TResponse> {
			private readonly IAsyncStreamReader<TResponse> _inner;

			public AsyncStreamReader(IAsyncStreamReader<TResponse> inner) {
				if (inner == null) throw new ArgumentNullException(nameof(inner));
				_inner = inner;
			}

			public async Task<bool> MoveNext(CancellationToken cancellationToken) {
				try {
					return await _inner.MoveNext(cancellationToken).ConfigureAwait(false);
				} catch (RpcException ex) {
					throw ConvertRpcException(ex);
				}
			}

			public TResponse Current => _inner.Current;
		}
	}
```
`ConvertRpcException` just pattern matches on `Trailers` and returns our typed exception. When not adding this interceptor to the channel, the correct `RpcException` comes back. I also verified that the server is returning the correct trailers. What am I doing wrong here?

**Comment by thefringeninja** [comment]
Upon further investigation is has nothing to do with the `Interceptor`, not sure where that came from! However it _is_ consistently happening on `AsyncServerStreamingCall`. 

Looking here: https://github.com/grpc/grpc-dotnet/blob/v2.26.0/src/Grpc.Net.Client/Internal/GrpcCall.cs#L440 it isn't clear how the exception gets notified back to the client.

**Comment by JamesNK** [comment]
It is hard to say what is going on without logging.

Enable logging on the server and the client - https://docs.microsoft.com/en-us/aspnet/core/grpc/diagnostics?view=aspnetcore-3.1 - and see what the cause of the cancellation results are.

**Comment by JunTaoLuo** [comment]
@thefringeninja, were you able to capture additional logs?

**Comment by thefringeninja** [comment]
@JunTaoLuo yes:
```
2020-01-29 13:14:30.022 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] SYSTEM INIT...
2020-01-29 13:14:30.029 [INF] [EventStore.Transport.Tcp.TcpServerListener] Starting "Normal" TCP listening on TCP endpoint: "127.0.0.1:1113".
2020-01-29 13:14:30.031 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] Service '"StorageReader"' initialized.
2020-01-29 13:14:30.031 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] Service '"StorageWriter"' initialized.
2020-01-29 13:14:30.032 [INF] [EventStore.Core.Services.Storage.ReaderIndex.IndexCommitter] TableIndex initialization...
2020-01-29 13:14:30.034 [INF] [EventStore.Core.Services.Storage.ReaderIndex.IndexCommitter] ReadIndex building...
2020-01-29 13:14:30.035 [DBG] [EventStore.Core.Services.Storage.ReaderIndex.IndexCommitter] ReadIndex rebuilding done: total processed 0 records, time elapsed: 00:00:00.0005564.
2020-01-29 13:14:30.044 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] Service '"StorageChaser"' initialized.
2020-01-29 13:14:30.045 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] SYSTEM START...
2020-01-29 13:14:30.046 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] IS UNKNOWN...
2020-01-29 13:14:30.047 [DBG] [EventStore.Core.Services.PersistentSubscription.PersistentSubscriptionService] Persistent subscriptions received state change to Unknown. Stopping listening
2020-01-29 13:14:30.047 [DBG] [EventStore.Core.Services.ElectionsService] ELECTIONS: STARTING ELECTIONS.
2020-01-29 13:14:30.048 [DBG] [EventStore.Core.Services.ElectionsService] ELECTIONS: (V=0) SHIFT TO LEADER ELECTION.
2020-01-29 13:14:30.048 [DBG] [EventStore.Core.Services.ElectionsService] ELECTIONS: (V=0) VIEWCHANGE FROM ["127.0.0.1:2112", {3474b49a-0230-47bf-b8eb-bf60a7dea9e5}].
2020-01-29 13:14:30.049 [DBG] [EventStore.Core.Services.ElectionsService] ELECTIONS: (V=0) MAJORITY OF VIEWCHANGE.
2020-01-29 13:14:30.050 [DBG] [EventStore.Core.Services.ElectionsService] ELECTIONS: (V=0) SHIFT TO PREPARE PHASE.
2020-01-29 13:14:30.050 [DBG] [EventStore.Core.Services.ElectionsService] ELECTIONS: (V=0) PREPARE_OK FROM "[127.0.0.1:2112,{3474b49a-0230-47bf-b8eb-bf60a7dea9e5}](L=-1,W=0,C=0,E-1@-1:{00000000-0000-0000-0000-000000000000},Priority=0)".
2020-01-29 13:14:30.050 [DBG] [EventStore.Core.Services.ElectionsService] ELECTIONS: (V=0) SHIFT TO REG_LEADER.
2020-01-29 13:14:30.054 [DBG] [EventStore.Core.Services.ElectionsService] ELECTIONS: (V=0) SENDING PROPOSAL CANDIDATE: "[127.0.0.1:2112,{3474b49a-0230-47bf-b8eb-bf60a7dea9e5}](L=-1,W=0,C=0,E-1@-1:{00000000-0000-0000-0000-000000000000},Priority=0)", ME: "[127.0.0.1:2112,{3474b49a-0230-47bf-b8eb-bf60a7dea9e5}](L=-1,W=0,C=0,E-1@-1:{00000000-0000-0000-0000-000000000000},Priority=0)".
2020-01-29 13:14:30.054 [DBG] [EventStore.Core.Services.ElectionsService] ELECTIONS: (V=0) ACCEPT FROM ["127.0.0.1:2112",{3474b49a-0230-47bf-b8eb-bf60a7dea9e5}] M=["127.0.0.1:2112",{3474b49a-0230-47bf-b8eb-bf60a7dea9e5}]).
2020-01-29 13:14:30.054 [INF] [EventStore.Core.Services.ElectionsService] ELECTIONS: (V=0) DONE. ELECTED MASTER = "[127.0.0.1:2112,{3474b49a-0230-47bf-b8eb-bf60a7dea9e5}](L=-1,W=0,C=0,E-1@-1:{00000000-0000-0000-0000-000000000000},Priority=0)". ME="[127.0.0.1:2112,{3474b49a-0230-47bf-b8eb-bf60a7dea9e5}](L=-1,W=0,C=0,E-1@-1:{00000000-0000-0000-0000-000000000000},Priority=0)".
2020-01-29 13:14:30.056 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] PRE-MASTER STATE, WAITING FOR CHASER TO CATCH UP...
2020-01-29 13:14:30.056 [DBG] [EventStore.Core.Services.PersistentSubscription.PersistentSubscriptionService] Persistent subscriptions received state change to PreMaster. Stopping listening
2020-01-29 13:14:30.057 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] IS MASTER... SPARTA!
2020-01-29 13:14:30.057 [DBG] [EventStore.Core.Services.PersistentSubscription.PersistentSubscriptionService] Persistent subscriptions Became Master so now handling subscriptions
2020-01-29 13:14:30.058 [DBG] [EventStore.Core.Services.Storage.StorageScavenger] Searching for incomplete scavenges on node "127.0.0.1:2113".
2020-01-29 13:14:30.064 [DBG] [EventStore.Core.Services.Storage.StorageScavenger] Setting max age for the "$scavenges" stream to 30.00:00:00.
2020-01-29 13:14:30.064 [DBG] [EventStore.Core.Services.Storage.StorageScavenger] Setting $ops read permission for the "$scavenges" stream
2020-01-29 13:14:30.067 [DBG] [EventStore.Core.Services.Storage.StorageScavenger] No incomplete scavenges found on node "127.0.0.1:2113".
2020-01-29 13:14:30.071 [DBG] [EventStore.Core.Services.Storage.EpochManager.EpochManager] === Writing E0@0:{f84264e1-6b65-4d92-9518-24ad0dc1840a} (previous epoch at -1).
2020-01-29 13:14:30.071 [DBG] [EventStore.Core.Services.Storage.EpochManager.EpochManager] === Update Last Epoch E0@0:{f84264e1-6b65-4d92-9518-24ad0dc1840a} (previous epoch at -1).
2020-01-29 13:14:30.105 [VRB] [EventStore.Core.Services.Monitoring.MonitoringService] Created stats stream '"$stats-127.0.0.1:2113"', code = Success
2020-01-29 13:14:30.124 [INF] [EventStore.Core.Services.UserManagement.UserManagementService] 'ops' user account has been created.
2020-01-29 13:14:30.124 [INF] [EventStore.Core.Services.UserManagement.UserManagementService] 'admin' user account has been created.
2020-01-29 13:14:30.124 [INF] [EventStore.Core.Services.UserManagement.UserManagementService] 'ops' user added to $users.
2020-01-29 13:14:30.125 [INF] [EventStore.Core.Services.UserManagement.UserManagementService] 'admin' user added to $users.
2020-01-29 13:14:30.127 [INF] [EventStore.Projections.Core.ProjectionsSubsystem] PROJECTIONS SUBSYSTEM: Starting components for Instance: 87633f11-a8eb-4e2b-a59b-a02d703a87ba
2020-01-29 13:14:30.127 [DBG] [EventStore.Projections.Core.Services.Management.ProjectionCoreCoordinator] PROJECTIONS: Projection Core Coordinator component starting. Correlation: 87633f11-a8eb-4e2b-a59b-a02d703a87ba
2020-01-29 13:14:30.127 [DBG] [EventStore.Projections.Core.Services.Management.ProjectionCoreCoordinator] PROJECTIONS: Starting Projections Core Coordinator
2020-01-29 13:14:30.129 [DBG] [EventStore.Projections.Core.Services.Management.ProjectionCoreCoordinator] PROJECTIONS: SubComponent Started: "EventReaderCoreService"
2020-01-29 13:14:30.129 [DBG] [EventStore.Projections.Core.ProjectionsSubsystem] PROJECTIONS SUBSYSTEM: Component '"ProjectionCoreCoordinator"' started for Instance: 87633f11-a8eb-4e2b-a59b-a02d703a87ba
2020-01-29 13:14:30.129 [INF] [EventStore.Projections.Core.ProjectionsSubsystem] PROJECTIONS SUBSYSTEM: All components started for Instance: 87633f11-a8eb-4e2b-a59b-a02d703a87ba
2020-01-29 13:14:30.129 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] Sub System '"Projections"' initialized.
2020-01-29 13:14:30.164 [DBG] [Grpc.Net.Client.Internal.GrpcCall] Starting gRPC call. Method type: 'ServerStreaming', URI: 'https://localhost:2113/event_store.client.streams.Streams/Read'.
2020-01-29 13:14:30.174 [DBG] [Grpc.Net.Client.Internal.GrpcCall] Sending message.
2020-01-29 13:14:30.179 [VRB] [Grpc.Net.Client.Internal.GrpcCall] Serialized '"EventStore.Client.Streams.ReadReq"' to 58 byte message.
2020-01-29 13:14:30.180 [VRB] [Grpc.Net.Client.Internal.GrpcCall] Message sent.
2020-01-29 13:14:30.191 [INF] [Microsoft.AspNetCore.Hosting.Diagnostics] Request starting HTTP/2 POST https://localhost:2113/event_store.client.streams.Streams/Read application/grpc 
2020-01-29 13:14:30.210 [DBG] [Microsoft.AspNetCore.Routing.Matching.DfaMatcher] 2 candidate(s) found for the request path '"/event_store.client.streams.Streams/Read"'
2020-01-29 13:14:30.212 [DBG] [Microsoft.AspNetCore.Routing.Matching.DfaMatcher] Endpoint '"gRPC - /event_store.client.streams.Streams/Read"' with route pattern '"/event_store.client.streams.Streams/Read"' is valid for the request path '"/event_store.client.streams.Streams/Read"'
2020-01-29 13:14:30.212 [DBG] [Microsoft.AspNetCore.Routing.Matching.DfaMatcher] Endpoint '"gRPC - gRPC - Unimplemented method for event_store.client.streams.Streams"' with route pattern '"event_store.client.streams.Streams/{unimplementedMethod}"' is valid for the request path '"/event_store.client.streams.Streams/Read"'
2020-01-29 13:14:30.213 [DBG] [Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware] Request matched endpoint '"gRPC - /event_store.client.streams.Streams/Read"'
2020-01-29 13:14:30.213 [INF] [Microsoft.AspNetCore.Routing.EndpointMiddleware] Executing endpoint '"gRPC - /event_store.client.streams.Streams/Read"'
2020-01-29 13:14:30.222 [DBG] [Grpc.AspNetCore.Server.ServerCallHandler] Reading message.
2020-01-29 13:14:30.225 [VRB] [Grpc.AspNetCore.Server.ServerCallHandler] Deserializing 58 byte message to '"EventStore.Client.Streams.ReadReq"'.
2020-01-29 13:14:30.230 [VRB] [Grpc.AspNetCore.Server.ServerCallHandler] Received message.
2020-01-29 13:14:30.241 [DBG] [EventStore.Client.EventStoreClient] Tombstoning stream "subscribe_to_stream_live_catches_deletions".
2020-01-29 13:14:30.243 [INF] [EventStore.Core.Services.Transport.Grpc.Enumerators+StreamSubscription] Live subscription 5f2631a7-9b11-41a6-8adb-eabf638aaa7c to "subscribe_to_stream_live_catches_deletions" running...
2020-01-29 13:14:30.243 [DBG] [Grpc.Net.Client.Internal.GrpcCall] Starting gRPC call. Method type: 'Unary', URI: 'https://localhost:2113/event_store.client.streams.Streams/Tombstone'.
2020-01-29 13:14:30.243 [DBG] [Grpc.Net.Client.Internal.GrpcCall] Sending message.
2020-01-29 13:14:30.244 [VRB] [Grpc.Net.Client.Internal.GrpcCall] Serialized '"EventStore.Client.Streams.TombstoneReq"' to 48 byte message.
2020-01-29 13:14:30.244 [VRB] [Grpc.Net.Client.Internal.GrpcCall] Message sent.
2020-01-29 13:14:30.244 [INF] [Microsoft.AspNetCore.Hosting.Diagnostics] Request starting HTTP/2 POST https://localhost:2113/event_store.client.streams.Streams/Tombstone application/grpc 
2020-01-29 13:14:30.245 [VRB] [EventStore.Core.Services.Transport.Grpc.Enumerators+StreamSubscription] Live subscription 5f2631a7-9b11-41a6-8adb-eabf638aaa7c to "subscribe_to_stream_live_catches_deletions" confirmed at "18446744073709551615".
2020-01-29 13:14:30.245 [DBG] [Microsoft.AspNetCore.Routing.Matching.DfaMatcher] 2 candidate(s) found for the request path '"/event_store.client.streams.Streams/Tombstone"'
2020-01-29 13:14:30.245 [DBG] [Microsoft.AspNetCore.Routing.Matching.DfaMatcher] Endpoint '"gRPC - /event_store.client.streams.Streams/Tombstone"' with route pattern '"/event_store.client.streams.Streams/Tombstone"' is valid for the request path '"/event_store.client.streams.Streams/Tombstone"'
2020-01-29 13:14:30.245 [DBG] [Microsoft.AspNetCore.Routing.Matching.DfaMatcher] Endpoint '"gRPC - gRPC - Unimplemented method for event_store.client.streams.Streams"' with route pattern '"event_store.client.streams.Streams/{unimplementedMethod}"' is valid for the request path '"/event_store.client.streams.Streams/Tombstone"'
2020-01-29 13:14:30.245 [DBG] [Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware] Request matched endpoint '"gRPC - /event_store.client.streams.Streams/Tombstone"'
2020-01-29 13:14:30.245 [INF] [Microsoft.AspNetCore.Routing.EndpointMiddleware] Executing endpoint '"gRPC - /event_store.client.streams.Streams/Tombstone"'
2020-01-29 13:14:30.247 [DBG] [Grpc.AspNetCore.Server.ServerCallHandler] Reading message.
2020-01-29 13:14:30.248 [VRB] [Grpc.AspNetCore.Server.ServerCallHandler] Deserializing 48 byte message to '"EventStore.Client.Streams.TombstoneReq"'.
2020-01-29 13:14:30.249 [VRB] [Grpc.AspNetCore.Server.ServerCallHandler] Received message.
2020-01-29 13:14:30.254 [VRB] [EventStore.Core.Services.Transport.Grpc.Enumerators+StreamSubscription] Live subscription 5f2631a7-9b11-41a6-8adb-eabf638aaa7c to "subscribe_to_stream_live_catches_deletions" loading any missed events starting at "9223372036854775807"
2020-01-29 13:14:30.257 [DBG] [Grpc.AspNetCore.Server.ServerCallHandler] Sending message.
2020-01-29 13:14:30.258 [VRB] [Grpc.AspNetCore.Server.ServerCallHandler] Serialized '"EventStore.Client.Streams.TombstoneResp"' to 8 byte message.
2020-01-29 13:14:30.259 [VRB] [Grpc.AspNetCore.Server.ServerCallHandler] Message sent.
2020-01-29 13:14:30.262 [INF] [Microsoft.AspNetCore.Routing.EndpointMiddleware] Executed endpoint '"gRPC - /event_store.client.streams.Streams/Tombstone"'
2020-01-29 13:14:30.265 [INF] [Microsoft.AspNetCore.Hosting.Diagnostics] Request finished in 20.8169ms 200 application/grpc
2020-01-29 13:14:30.267 [VRB] [Grpc.Net.Client.Internal.GrpcCall] Response headers received.
2020-01-29 13:14:30.269 [INF] [EventStore.Core.Services.Transport.Grpc.Enumerators+StreamSubscription] Live subscription 5f2631a7-9b11-41a6-8adb-eabf638aaa7c to "subscribe_to_stream_live_catches_deletions" disposed.
2020-01-29 13:14:30.274 [DBG] [Grpc.Net.Client.Internal.GrpcCall] Reading message.
2020-01-29 13:14:30.275 [VRB] [Grpc.Net.Client.Internal.GrpcCall] Deserializing 8 byte message to '"EventStore.Client.Streams.TombstoneResp"'.
2020-01-29 13:14:30.277 [VRB] [Grpc.Net.Client.Internal.GrpcCall] Received message.
2020-01-29 13:14:30.278 [DBG] [Grpc.Net.Client.Internal.GrpcCall] Finished gRPC call.
2020-01-29 13:14:30.272 [INF] [Grpc.AspNetCore.Server.ServerCallHandler] Error status code 'FailedPrecondition' raised.
Grpc.Core.RpcException: Status(StatusCode=FailedPrecondition, Detail="Event stream 'subscribe_to_stream_live_catches_deletions' is deleted.")
   at EventStore.Core.Services.Transport.Grpc.Enumerators.StreamSubscription.LiveStreamSubscription.MoveNextAsync() in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Enumerators.StreamSubscription.cs:line 446
   at EventStore.Core.Services.Transport.Grpc.Enumerators.StreamSubscription.MoveNextCoreAsync() in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Enumerators.StreamSubscription.cs:line 96
   at EventStore.Core.Services.Transport.Grpc.Enumerators.StreamSubscription.MoveNextAsync() in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Enumerators.StreamSubscription.cs:line 80
   at EventStore.Core.Services.Transport.Grpc.Streams.Read(ReadReq request, IServerStreamWriter`1 responseStream, ServerCallContext context) in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Streams.Read.cs:line 144
   at EventStore.Core.Services.Transport.Grpc.Streams.Read(ReadReq request, IServerStreamWriter`1 responseStream, ServerCallContext context) in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Streams.Read.cs:line 144
   at EventStore.Core.Services.Transport.Grpc.Streams.Read(ReadReq request, IServerStreamWriter`1 responseStream, ServerCallContext context) in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Streams.Read.cs:line 144
   at Grpc.AspNetCore.Server.Internal.CallHandlers.ServerStreamingServerCallHandler`3.HandleCallAsyncCore(HttpContext httpContext, HttpContextServerCallContext serverCallContext)
   at Grpc.AspNetCore.Server.Internal.CallHandlers.ServerStreamingServerCallHandler`3.HandleCallAsyncCore(HttpContext httpContext, HttpContextServerCallContext serverCallContext)
   at Grpc.AspNetCore.Server.Internal.CallHandlers.ServerCallHandlerBase`3.<HandleCallAsync>g__AwaitHandleCall|17_0(HttpContextServerCallContext serverCallContext, Method`2 method, Task handleCall)

2020-01-29 13:14:30.288 [INF] [Microsoft.AspNetCore.Routing.EndpointMiddleware] Executed endpoint '"gRPC - /event_store.client.streams.Streams/Read"'
2020-01-29 13:14:30.288 [INF] [Microsoft.AspNetCore.Hosting.Diagnostics] Request finished in 99.2145ms 200 application/grpc
2020-01-29 13:14:30.289 [VRB] [Grpc.Net.Client.Internal.GrpcCall] Response headers received.
2020-01-29 13:14:30.292 [ERR] [Grpc.Net.Client.Internal.GrpcCall] Call failed with gRPC error status. Status code: 'FailedPrecondition', Message: '"Event stream 'subscribe_to_stream_live_catches_deletions' is deleted."'.
2020-01-29 13:14:30.292 [DBG] [Grpc.Net.Client.Internal.GrpcCall] Finished gRPC call.
2020-01-29 13:14:30.302 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] IS SHUTTING DOWN...
2020-01-29 13:14:30.429 [VRB] [EventStore.Core.Bus.InMemoryBus] SLOW BUS MSG ["MainBus"]: "BecomeShuttingDown" - 126ms. Handler: "ClusterVNode".
2020-01-29 13:14:30.438 [VRB] [EventStore.Core.Bus.QueuedHandlerMRES] SLOW QUEUE MSG ["MainQueue"]: "RequestShutdown" - 135ms. Q: 0/5.
2020-01-29 13:14:30.438 [DBG] [EventStore.Core.Services.PersistentSubscription.PersistentSubscriptionService] Persistent subscriptions received state change to ShuttingDown. Stopping listening
2020-01-29 13:14:30.438 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] Service '"StorageWriter"' has shut down.
2020-01-29 13:14:30.438 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] Service '"StorageReader"' has shut down.
2020-01-29 13:14:30.438 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] Service '"Storage Chaser"' has shut down.
2020-01-29 13:14:30.438 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] Service '"HttpServer [127.0.0.1:2113]"' has shut down.
2020-01-29 13:14:30.453 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] Service '"ReplicationTrackingService"' has shut down.
2020-01-29 13:14:30.453 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] All Services Shutdown.
2020-01-29 13:14:30.455 [INF] [EventStore.Core.Services.VNode.ClusterVNodeController] ========== ["127.0.0.1:2112"] IS SHUT DOWN.
2020-01-29 13:14:30.456 [DBG] [Microsoft.AspNetCore.Hosting.Diagnostics] Hosting shutdown
```
The funny thing is, this is pretty much what I expect to see - this is the correct exception. It appears as if this is getting ignored somewhere in the grpc client.

If you want, I can filter out our internal logs and re-paste.

**Comment by thefringeninja** [comment]
Ok, found some more information. After getting logging talking to xunit's ITestOutputHelper correctly, and setting `ThrowOperationCanceledOnCancellation = false`, I now see the following when we get gRPC cancelled:
```
System.OperationCanceledException: Flush was canceled on underlying PipeWriter.
   at System.IO.Pipelines.ThrowHelper.ThrowOperationCanceledException_FlushCanceled()
   at System.IO.Pipelines.PipeWriterStream.GetFlushResultAsTask(ValueTask`1 valueTask)
   at System.IO.Pipelines.PipeWriterStream.WriteAsync(ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at Grpc.Net.Client.StreamExtensions.WriteHeaderAsync(Stream stream, Int32 length, Boolean compress, CancellationToken cancellationToken)
   at Grpc.Net.Client.StreamExtensions.WriteMessageAsync[TMessage](Stream stream, ILogger logger, TMessage message, Action`2 serializer, String grpcEncoding, Nullable`1 maximumMessageSize, Dictionary`2 compressionProviders, CallOptions callOptions)
```
We of course are not explicitly cancelling anything, as far as I can tell.

**Comment by JunTaoLuo** [comment]
There's an exception that's not handled on the server:
```
Grpc.Core.RpcException: Status(StatusCode=FailedPrecondition, Detail="Event stream 'subscribe_to_stream_live_catches_deletions' is deleted.")
   at EventStore.Core.Services.Transport.Grpc.Enumerators.StreamSubscription.LiveStreamSubscription.MoveNextAsync() in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Enumerators.StreamSubscription.cs:line 446
   at EventStore.Core.Services.Transport.Grpc.Enumerators.StreamSubscription.MoveNextCoreAsync() in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Enumerators.StreamSubscription.cs:line 96
   at EventStore.Core.Services.Transport.Grpc.Enumerators.StreamSubscription.MoveNextAsync() in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Enumerators.StreamSubscription.cs:line 80
   at EventStore.Core.Services.Transport.Grpc.Streams.Read(ReadReq request, IServerStreamWriter`1 responseStream, ServerCallContext context) in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Streams.Read.cs:line 144
   at EventStore.Core.Services.Transport.Grpc.Streams.Read(ReadReq request, IServerStreamWriter`1 responseStream, ServerCallContext context) in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Streams.Read.cs:line 144
   at EventStore.Core.Services.Transport.Grpc.Streams.Read(ReadReq request, IServerStreamWriter`1 responseStream, ServerCallContext context) in /home/jwow/dev/EventStore/src/EventStore.Core/Services/Transport/Grpc/Streams.Read.cs:line 144
   at Grpc.AspNetCore.Server.Internal.CallHandlers.ServerStreamingServerCallHandler`3.HandleCallAsyncCore(HttpContext httpContext, HttpContextServerCallContext serverCallContext)
   at Grpc.AspNetCore.Server.Internal.CallHandlers.ServerStreamingServerCallHandler`3.HandleCallAsyncCore(HttpContext httpContext, HttpContextServerCallContext serverCallContext)
   at Grpc.AspNetCore.Server.Internal.CallHandlers.ServerCallHandlerBase`3.<HandleCallAsync>g__AwaitHandleCall|17_0(HttpContextServerCallContext serverCallContext, Method`2 method, Task handleCall)
```

From the first set of logs, I don't see a `TaskCancelledException` though. It seems the client received the error status correctly:
```
2020-01-29 13:14:30.292 [ERR] [Grpc.Net.Client.Internal.GrpcCall] Call failed with gRPC error status. Status code: 'FailedPrecondition', Message: '"Event stream 'subscribe_to_stream_live_catches_deletions' is deleted."'.
2020-01-29 13:14:30.292 [DBG] [Grpc.Net.Client.Internal.GrpcCall] Finished gRPC call.
```

**Comment by thefringeninja** [comment]
This is the expected behavior:

https://github.com/thefringeninja/EventStore/blob/read-sanity/src/EventStore.Client.Tests/Streams/subscribe_to_stream_live.cs#L127

**Comment by JunTaoLuo** [comment]
You might be running into the issue we encountered in https://github.com/grpc/grpc-dotnet/issues/723. Can you try out the nightly at https://github.com/grpc/grpc-dotnet#grpc-nuget-feed>

**Comment by JamesNK** [comment]
A nightly isn't required. 2.27.0 on NuGet fixes that issue.

**Comment by anurse** [comment]
Closing as this is fixed in 2.27.0
